### PR TITLE
Rush now automatically installs the "rush-lib" version that was requested in rush.json

### PR DESCRIPTION
**Pull request by iclanton**
This PR addresses two major issues that commonly affect new releases of Rush:

- **Every developer should ideally be using a consistent version of Rush**:  Otherwise, problems can be difficult to diagnose, and people can be tripped up by bugs that were already fixed in the latest version.  The existing `rushMinimumVersion` setting isn't a very good solution, because it can quickly become outdated, and it's annoying for developers to reinstall Rush every time it changes.

- **The repo maintainer should choose when to upgrade**:  Sometimes new releases of Rush require some migration steps before they can be used, or they introduce regressions that require a rollback while a fix is being prepared.  Asking everyone to manually downgrade their Rush version is a bad experience.

This PR enables the repo maintainer to specify the desired Rush release in their `rush.json` file.  Whenever someone invokes the Rush tool in that folder, the requested release will be automatically installed and used, without any special actions by the developer.

The idea is to move all the Rush implementation into **@microsoft/rush-lib**, so that the **@microsoft/rush** package becomes just a launcher.  It reads just the `rushVersion` field from `rush.json` and then installs the appropriate **@microsoft/rush-lib** release in a system-wide cache in the user's home directory.  It then passes control to **@microsoft/rush-lib**, which determines the config file format and even command-line experience.  (If invoked outside of a repo folder, then the globally installed **@microsoft/rush-lib** is used.)

### CODE REVIEW

**Comment by iclanton** [comment]
> 

Move to root and call this start.ts

---
Refers to: apps/rush-lib/src/cli/executeCli.ts:18 in ca7b36b. [](commit_id = ca7b36b0fbe983e28632c43708f73d592c284343, deletion_comment = False)

File: apps/rush/src/rush.ts [hunk 1]
```diff
@@ -1,15 +1,13 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
```
Comment (iclanton) [inline apps/rush/src/rush.ts]: rename to start.ts

File: apps/rush-lib/src/cli/executeCli.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as os from 'os';
+import * as colors from 'colors';
+
+import rushVersion from '../rushVersion';
+import RushCommandLineParser from './actions/RushCommandLineParser';
+
+export function executeCli(): void {
```
Comment (iclanton) [inline apps/rush-lib/src/cli/executeCli.ts]: >executeCli [](start = 16, length = 10)

Have this take the version of the c2r wrapper to future-proof.

File: apps/rush/src/RushVersionManager.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+export default class RushVersionManager {
+  private _rushDirectory: string;
+
+  constructor(homeDirectory: string) {
+    this._rushDirectory = path.join(homeDirectory, '.rush');
+  }
+
+  public ensureRushVersionInstalled(version: string): RushWrapper {
+    const isLegacyRushVersion: boolean = semver.lt(version, RUSH_TRANSITIONAL_VERSION);
```
Comment (iclanton) [inline apps/rush/src/RushVersionManager.ts]: >const isLegacyRushVersion: boolean = semver.lt(version, RUSH_TRANSITIONAL_VERSION); [](start = 4, length = 83)

make sure this works with 1.4.0

File: apps/rush/src/RushVersionManager.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    } else {
+      return new RushWrapper(() => {
+        const rushCliEntrypoint: IRushCliEntrypoint = require(path.join(
+          expectedRushPath,
+          'node_modules',
+          '@microsoft',
+          'rush-lib',
+          'lib',
+          'cli',
+          'executeCli'
```
Comment (iclanton) [inline apps/rush/src/RushVersionManager.ts]: rush-lib/start

File: apps/rush/src/MinimalRushConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const rushJsonLocation: string = RushConfiguration.findRushJsonLocation();
+    return MinimalRushConfiguration.loadFromConfigurationFile(rushJsonLocation);
+  }
+
+  public static loadFromConfigurationFile(rushJsonFilename: string): MinimalRushConfiguration {
+    try {
+      const minimalRushConfigurationJson: IMinimalRushConfigurationJson = JsonFile.load(rushJsonFilename);
+      return new MinimalRushConfiguration(minimalRushConfigurationJson);
+    } catch (e) {
+      throw new Error(`Failed to load rush.json: ${e}`);
```
Comment (iclanton) [inline apps/rush/src/MinimalRushConfiguration.ts]: >throw new Error(`Failed to load rush.json: ${e}`); [](start = 6, length = 50)

Don't throw here, just call the bundled rush-lib.

File: apps/rush-lib/src/cli/executeCli.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  console.log(os.EOL + colors.bold(`Rush Multi-Package Build Tool ${rushVersion}`)
```
Comment (iclanton) [inline apps/rush-lib/src/cli/executeCli.ts]: >  [](start = 65, length = 1)

+ Engine

Comment (iclanton) [inline apps/rush-lib/src/cli/executeCli.ts]: >` [](start = 80, length = 1)

+ (unmanaged)

**Comment by pgonzal** [comment]
@dzearing @christiango  FYI -- This is a pretty exciting feature for Rush; it should greatly reduce the risk of getting broken by a Rush update.

File: apps/rush/src/MinimalRushConfiguration.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+import { JsonFile } from '@microsoft/node-core-library';
+import { RushConfiguration } from '@microsoft/rush-lib';
+
+interface IMinimalRushConfigurationJson {
+  rushMinimumVersion: string;
+  rushVersion?: string;
+}
+
+export default class MinimalRushConfiguration {
```
Comment (pgonzal) [inline apps/rush/src/MinimalRushConfiguration.ts]: >export [](start = 0, length = 6)

Can you add a class comment explaining what this is for

File: apps/rush-lib/src/data/RushConfiguration.ts [hunk 8]
```diff
@@ -186,10 +199,25 @@ export default class RushConfiguration {
       }
       currentFolder = parentFolder;
     }
+
     throw new Error('Unable to find rush.json configuration file');
```
Comment (pgonzal) [inline apps/rush-lib/src/data/RushConfiguration.ts]: >throw new Error('Unable to find rush.json configuration file'); [](start = 4, length = 63)

If Rush is invoked outside of a repo folder, then there will not be any "rush.json" in any parent folder.  This is an expected situation.  If the code needs to test that, it should NOT have to do that by catching an unexceptional exception; instead, we should change the API to e.g. tryFindRushJsonLocation() which returns undefined if the path cannot be found.

File: apps/rush/src/MinimalRushConfiguration.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      return undefined;
+    }
+  }
+
+  private constructor(minimalRushConfigurationJson: IMinimalRushConfigurationJson) {
+    this._rushVersion = minimalRushConfigurationJson.rushVersion || minimalRushConfigurationJson.rushMinimumVersion;
+    this._homeFolder = RushConfiguration.getHomeDirectory();
+  }
+
+  public get rushVersion(): string {
```
Comment (pgonzal) [inline apps/rush/src/MinimalRushConfiguration.ts]: >public [](start = 2, length = 6)

Document what this is, and what an example value looks like

File: apps/rush/src/MinimalRushConfiguration.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    return this._rushVersion;
+  }
+
+  public get homeFolder(): string {
```
Comment (pgonzal) [inline apps/rush/src/MinimalRushConfiguration.ts]: >public [](start = 2, length = 6)

Document what this is, and what an example value looks like (e.g. is it always an absolute path? Can it ever be an empty string?)

File: apps/rush/src/RushVersionManager.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import * as path from 'path';
+import * as fsx from 'fs-extra';
+import * as semver from 'semver';
+
+import { Utilities } from '@microsoft/rush-lib';
+import * as rushLibCli from '@microsoft/rush-lib/lib/start';
+
+import RushWrapper from './RushWrapper';
+
+const RUSH_TRANSITIONAL_VERSION: string = '4.0.0';
```
Comment (pgonzal) [inline apps/rush/src/RushVersionManager.ts]: >const [](start = 0, length = 5)

Add a comment -- what is significant about version 4?

File: apps/rush/src/RushVersionManager.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+const MAX_INSTALL_ATTEMPTS: number = 5;
```
Comment (pgonzal) [inline apps/rush/src/RushVersionManager.ts]: >5; [](start = 37, length = 2)

This seems like a lot of retries for something that (1) is usually manually invoked and (2) is generally a cache hit.

Maybe 3?

File: apps/rush/src/RushVersionManager.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    this._currentPackageVersion = currentPackageVersion;
+  }
+
+  public ensureRushVersionInstalled(version: string): RushWrapper {
+    const isLegacyRushVersion: boolean = semver.lt(version, RUSH_TRANSITIONAL_VERSION);
+    const expectedRushPath: string = path.join(this._rushDirectory, `rush-${version}`);
+    const expectedRushInstalledFlagPath: string = path.join(expectedRushPath, 'last-install.flag');
+    if (!fsx.existsSync(expectedRushInstalledFlagPath)) {
+      // Need to install Rush
+      console.log(`Rush version ${version} is not currently installed. Installing...`);
```
Comment (pgonzal) [inline apps/rush/src/RushVersionManager.ts]: >console.log(`Rush version ${version} is not currently installed. Installing...`); [](start = 6, length = 81)

If two different CI jobs try to do this simultaneously, can something bad happen?  e.g. could we end up with a corrupted folder?

How did InstallRushOnlyIfNeeded.js handle this?

**Comment by pgonzal** [comment]
> }

Your PR greatly reduces rush's contract with rush-lib. Before your major version bump, let's come back in a separate PR and remove a bunch of this junk from the public API

---
Refers to: common/reviews/api/rush-lib.api.ts:418 in e3000ce. [](commit_id = e3000cec96357b103593a5be50f35c2fbceb2cda, deletion_comment = False)

File: apps/rush-lib/src/rush.schema.json [hunk 14]
```diff
@@ -14,8 +14,8 @@
       "type": "string",
       "pattern": "^[0-9]+\\.[0-9]+\\.[0-9]+$"
     },
-    "rushMinimumVersion": {
-      "description": "The minimum version of the Rush tool that must be installed in order to build this repository.",
+    "rushVersion": {
+      "description": "The version of the Rush tool that must be installed in order to build this repository.",
```
Comment (pgonzal) [inline apps/rush-lib/src/rush.schema.json]: >that must be installed [](start = 51, length = 22)

"that must be installed in order to" --> "that will be used to"

File: apps/rush-lib/src/start.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+import * as colors from 'colors';
+
+import rushVersion from './rushVersion';
+import RushCommandLineParser from './cli/actions/RushCommandLineParser';
+
+// tslint:disable-next-line:export-name
+export function executeCli(wrapperVersion: string, isManaged: boolean = false): void {
+  console.log(
+    EOL +
+    colors.bold(`Rush Multi-Package Build Tool ${rushVersion}` + colors.yellow(isManaged ? '' : '(unmanaged)')) +
```
Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: >unmanaged [](start = 98, length = 9)

Add a little code comment explaining that "(unmanaged)" is used to make it easier to recognize repos that don't lock down their Rush version, or cases where people invoke the tool outside of a repo.  It's mainly to help us when we see screenshots or log files

File: apps/rush-lib/src/start.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { EOL } from 'os';
+import * as colors from 'colors';
+
+import rushVersion from './rushVersion';
+import RushCommandLineParser from './cli/actions/RushCommandLineParser';
+
+// tslint:disable-next-line:export-name
```
Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: >// [](start = 0, length = 2)

Add a detailed JSDoc comment explaining that this is the main entry point for rush invoking rush-lib, and document what the parameters are and why we need them.

File: apps/rush-lib/src/start.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+export function executeCli(wrapperVersion: string, isManaged: boolean = false): void {
```
Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: > isManaged: boolean = false [](start = 50, length = 27)

This is a formal plug-in contract; nothing should be optional

Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: >wrapperVersion [](start = 27, length = 14)

"wrapper" seems like the wrong term.  I think I would just call this "rushVersion" or "callerVersion" (or maybe "launcherVersion"?)

Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: >executeCli [](start = 16, length = 10)

since the file is start.ts, should the function be called "start"?

Comment (iclanton) [reply apps/rush-lib/src/start.ts]: I'd like to limit the number of arguments, but I'll make the second one non-optional.

---
In reply to: [145844452](https://github.com/Microsoft/web-build-tools/pull/397#discussion_r145844452) [](ancestors = 145844452)

**Comment by iclanton** [comment]
> }

Yeah we should.

---
In reply to: [338059894](https://github.com/Microsoft/web-build-tools/pull/397#issuecomment-338059894) [](ancestors = 338059894)

---
Refers to: common/reviews/api/rush-lib.api.ts:418 in e3000ce. [](commit_id = e3000cec96357b103593a5be50f35c2fbceb2cda, deletion_comment = False)

File: apps/rush-lib/src/start.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+// See LICENSE in the project root for license information.
+
+import { EOL } from 'os';
+import * as colors from 'colors';
+
+import rushVersion from './rushVersion';
+import RushCommandLineParser from './cli/actions/RushCommandLineParser';
+
+/**
+ * Executes the Rush CLI. This is expected to be called by the @microsoft/rush package, which acts as a version manager
```
Comment (pgonzal) [inline apps/rush-lib/src/start.ts]: >@ [](start = 63, length = 1)

Use a backslash so this isn't interpreted as a JSDoc tag

**Pull request merged**